bee-network/src/swarm/protocols/gossip/gossip_io.rs

```diff
@@ -0,0 +1,149 @@
+pub type GossipSender = mpsc::UnboundedSender<Vec<u8>>;
+
+/// A shorthand for an unbounded channel receiver.
+pub type GossipReceiver = UnboundedReceiverStream<Vec<u8>>;
+
+pub fn gossip_channel() -> (GossipSender, GossipReceiver) {
+    let (sender, receiver) = mpsc::unbounded_channel();
+    (sender, UnboundedReceiverStream::new(receiver))
+}
+
+pub fn spawn_gossip_in_processor(
+    peer_id: PeerId,
+    mut reader: ReadHalf<NegotiatedSubstream>,
+    incoming_gossip_sender: GossipSender,
+    internal_event_sender: InternalEventSender,
+) {
+    tokio::spawn(async move {
+        let mut msg_buf = vec![0u8; MSG_BUFFER_SIZE];
+        let mut msg_len = 0;
+
+        loop {
+            if recv_valid_message(&mut reader, &mut msg_buf, &mut msg_len).await {
+                if incoming_gossip_sender.send(msg_buf[..msg_len].to_vec()).is_err() {
+                    debug!("gossip-in: receiver dropped locally.");
+
+                    // The receiver of this channel was dropped, maybe due to a shutdown. There is nothing we can do to
+                    // salvage this situation, hence we drop the connection.
+                    break;
+                }
+            } else {
+                debug!("gossip-in: stream closed remotely.");
+
+                // NB: The network service will not shut down before it has received the `ProtocolDropped` event from
+                // all once connected peers, hence if the following send fails, then it must be
+                // considered a bug.
+
+                // The remote peer dropped the connection.
+                internal_event_sender
+                    .send(InternalEvent::ProtocolDropped { peer_id })
+                    .expect("The service must not shutdown as long as there are gossip tasks running.");
+
+                break;
+            }
+        }
+
+        // Reasons why this task might end:
+        // (1) The remote dropped the TCP connection.
+        // (2) The local dropped the gossip_in receiver channel.
+
+        debug!("gossip-in: exiting gossip-in processor for {}.", alias!(peer_id));
+    });
+}
+
+async fn recv_valid_message<S>(stream: &mut S, message: &mut [u8], message_len: &mut usize) -> bool
+where
+    S: AsyncRead + Unpin,
+{
+    if let Ok(msg_len) = stream.read(message).await {
+        if msg_len == 0 {
+            false
+        } else {
+            *message_len = msg_len;
+            true
+        }
+    } else {
+        false
+    }
+}
+
+pub fn spawn_gossip_out_processor(
+    peer_id: PeerId,
+    mut writer: WriteHalf<NegotiatedSubstream>,
+    outgoing_gossip_receiver: GossipReceiver,
+    internal_event_sender: InternalEventSender,
+) {
+    tokio::spawn(async move {
+        let mut outgoing_gossip_receiver = outgoing_gossip_receiver.fuse();
+
+        // If the gossip sender dropped we end the connection.
+        while let Some(message) = outgoing_gossip_receiver.next().await {
+            let msg_len = message.len();
```

Comment:
What's the reason for this realias?

Reply:
probably stems from old code where the message might have been moved. Can't remember. Removed that line.